Cython/Compiler/ExprNodes.py

```diff
@@ -11843,12 +11842,17 @@ def analyse_operation(self, env):
             self.coerce_operands_to_pyobjects(env)
             self.type = self.result_type(self.operand1.type,
                                          self.operand2.type, env)
-            assert self.type.is_pyobject
             self.is_temp = 1
+            if not self.type.is_pyobject:
+                original_type, self.type = self.type, py_object_type
+                # Hopefully this can be optimized out in some cases
+                return self.coerce_to(original_type, env)
+
         elif self.is_cpp_operation():
             self.analyse_cpp_operation(env)
         else:
             self.analyse_c_operation(env)
+        return self
```

Comment:
I think you forgot to change a few `analyse_operation` methods, at least the ones in `MemoryViewIndexNode` and `MemoryViewSliceNode`.

Reply:
My thought was that these methods weren't obvious related in that the binop nodes were never going to be calling the memory indexing methods. But I've changed them for consistency.